### PR TITLE
Provide aggregated validation errors for version and kind

### LINKED ISSUE #54098: Gather all errors before failing openapi validation
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**: When kind/group/version is not present, only one of the missing file is reported. They should all be reported as missing.

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration**:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by spzala**
Currently the validation checks is done individually for version and
kind group. For example, if user provided yaml file is missing apiVersion
and kind fields, first they will receive error on apiVersion. Once user
update the file and try to recreate, an error on missing kind is displayed.
The behavior is same for wrong types of the fields.
These errors should be aggregated and displayed.

Examples of current validation:
1
test.yaml is missing apiVersion and kind:
$kubectl.sh create -f /home/sahdev/go/src/bugfixes/test.yaml
error: error validating "/home/sahdev/go/src/bugfixes/test.yaml": error validating data: apiVersion not set; if you choose to ignore these errors, turn validation off with --validate=false
2.
test.yaml is fixed for apiVersion but missing kind:
$kubectl.sh create -f /home/sahdev/go/src/bugfixes/test.yaml
error: error validating "/home/sahdev/go/src/bugfixes/test.yaml": error validating data: kind not set; if you choose to ignore these errors, turn validation off with --validate=false

Examples with aggregated validation: 
1.
error: error validating "/home/sahdev/go/src/bugfixes/test.yaml": error validating data: [apiVersion not set, kind not set]; if you choose to ignore these errors, turn validation off with --validate=false
2.
error: error validating "/home/sahdev/go/src/bugfixes/testmix.yaml": error validating data: [apiVersion isn't string type, kind not set]; if you choose to ignore these errors, turn validation off with --validate=false

**What this PR does / why we need it**:
To provide aggregated validations to user for version and kind group.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #54098

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by spzala** [comment]
@pwittrock thanks!

**Review by apelisse — COMMENTED** [review]
I think that's a good idea, and I like how the code looks.

Do you think we could move the error aggregation up to `ValidateBytes()`, where other errors are also aggregated, and have `getObjectKind` return a list of errors? It might be more consistent.

Thank you for this change!

**Comment by spzala** [comment]
@apelisse hi and thanks so much for the comments. Sorry if I misunderstood your suggestion but actually I did initially try to modify the  ```ValidateBytes``` in the Schema but since ```getObjectKind``` currently ```return``` individual errors I found that it might be best choice to aggregate errors in the ```getObjectKind``` itself and leaving a single call to ```getObjectKind``` as such. Thanks!!

**Comment by apelisse** [comment]
What I meant is (and I of course may be wrong):
Change the signature of `getObjectKind`:
```
// Returns list of errors instead of single error
func getObjectKind(object interface{}) (schema.GroupVersionKind, []error)  
```

And change `ValidateBytes`:
```
gvk, errs := getObjectKind(obj)
if errs != nil {
	return utilerrors.NewAggregate(errs)
}
```

**Comment by spzala** [comment]
@apelisse thanks much!! Since ```getObjectKind``` is referenced at couple of places, I thought I wouldn't change the signature but since it's only two places so changing signature might not be a bad idea at all. I will check it out soon.

**Comment by apelisse** [comment]
Thank you for mentionning that it's called in 2 places. It actually reinforces my point:

If you look at that:
https://github.com/spzala/kubernetes/blob/daf6c57aec795ab9dc40077fbfc6ba99dab62f5c/pkg/kubectl/cmd/util/openapi/validation/validation.go#L66-L68

This would do nesting lists, which would be pretty ugly :-)

Thanks for working on that!

**Comment by spzala** [comment]
@apelisse :-) agree with you!! I will be updating my finding/changes later today. Thanks!!

**Comment by spzala** [comment]
@apelisse updated new commit :-). Thanks!!

File: pkg/kubectl/cmd/util/openapi/validation/validation.go
Comment (apelisse) [inline pkg/kubectl/cmd/util/openapi/validation/validation.go]: you have an extra space after `[]`

File: pkg/kubectl/cmd/util/openapi/validation/validation.go
Comment (apelisse) [inline pkg/kubectl/cmd/util/openapi/validation/validation.go]: maybe move that line somewhere else. It's kind of awkward to seperate `field := ...` and `if fields == nil {`

File: pkg/kubectl/cmd/util/openapi/validation/validation.go
Comment (apelisse) [inline pkg/kubectl/cmd/util/openapi/validation/validation.go]: Can we move this to the end? I know we can't because we don't know if `apiVersion` is a string. I'll explain in a different comment.

**Review by apelisse — COMMENTED** [review]
Thank you for your change. I think it can be further improved.

Maybe we should change the function further to:
- Fill in the GVK as we read more from the object,
- Fill listErrors as we fail to do so,
- Return both gvk and listErrors only once at the end (beside the first return statement if `fields` is nil)

File: pkg/kubectl/cmd/util/openapi/validation/validation.go
Comment (apelisse) [inline pkg/kubectl/cmd/util/openapi/validation/validation.go]: I forgot the most important: This is appending a nested list to the list.

It should read:
```
errs = append(errs, err...) // And maybe change the of errs?
```

Comment (spzala) [reply pkg/kubectl/cmd/util/openapi/validation/validation.go]: @apelisse I didn't know about variadic so thanks much for helping me educate here. I made changes and tested, it works :-)

File: pkg/kubectl/cmd/util/openapi/validation/validation.go
Comment (spzala) [reply pkg/kubectl/cmd/util/openapi/validation/validation.go]: Yup, makes sense!

File: pkg/kubectl/cmd/util/openapi/validation/validation.go
Comment (spzala) [reply pkg/kubectl/cmd/util/openapi/validation/validation.go]: @apelisse +1 and Yes, that's the reason I didn't move but I see that the need and found easier way of doing it. So I actually tried to take care of it in the new commit I just uploaded but please let me know if I didn't do it the way you are thinking :-). Thanks!

**Review by apelisse — APPROVED** [review]
That's great, thanks @spzala !

**Comment by apelisse** [comment]
Can you please squash these commits? Thank you

**Comment by spzala** [comment]
@apelisse yrw and thank you!! I will squash and also seems some script related errors so going to work on it to fix'em.

**Comment by spzala** [comment]
@apelisse hi, as you can see in the last commit I removed couple of extra spaces to fix go formatting errors but I noticed that bot has removed lgtm label :(. I hope I took of all formatting errors though. Thanks!

**Comment by apelisse** [comment]
yeah, you still have 4 commits, please squash. I'll be happy to re-lgtm once you've pushed!

Thanks

File: pkg/kubectl/cmd/util/openapi/validation/validation.go
Comment (shiywang) [inline pkg/kubectl/cmd/util/openapi/validation/validation.go]: I think this name could be `allErrors` `errors` or `errList`... also please squash your commit

File: pkg/kubectl/cmd/util/openapi/validation/validation.go
Comment (apelisse) [inline pkg/kubectl/cmd/util/openapi/validation/validation.go]: This should be `errs != nil`

Comment (apelisse) [reply pkg/kubectl/cmd/util/openapi/validation/validation.go]: It looks like the code that you're changing is not validated by any unit-test. Would you mind (in this PR or in a in follow-up PR) improve the unit-test thoroughness?. For example, I could have missed that mistake, and we should be confident that the tests are going to catch it.

**Review by apelisse — CHANGES_REQUESTED** [review]
Beside that little error, looks very good to me!

File: pkg/kubectl/cmd/util/openapi/validation/validation.go
Comment (spzala) [reply pkg/kubectl/cmd/util/openapi/validation/validation.go]: @apelisse hi, yup agree and absolutely I would work on it as a follow up PR(s). I actually wanted to do it and tried playing with validation_test. So I wanted to touch base with you on it too. Thanks!!

**Comment by spzala** [comment]
@apelisse great catch and yes totally agree on tests. I will be creating a follow up PR on it. Thanks so much!!

**Comment by apelisse** [comment]
/lgtm

Thank you @spzala

**Review by apelisse — APPROVED** [review]

**Comment by spzala** [comment]
@apelisse sorry about the formatting error, I have fixed it and thanks much for showing me the formatting script :) !!

**Pull request merged**
